compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -3324,6 +3325,13 @@ fn print_disambiguation_help<'tcx>(
         {
             let def_kind_descr = tcx.def_kind_descr(item.kind.as_def_kind(), item.def_id);
             let item_name = item.ident(tcx);
+            let first_arg_type = tcx
+                .fn_sig(item.def_id)
+                .instantiate_identity()
+                .skip_binder()
+                .inputs()
+                .get(0)
+                .and_then(|first| Some(first.peel_refs()));
```

Comment:
This makes sense but it's prone to errors:

```
trait A { fn f(self); }
trait B { fn f(self); }

#[derive(Clone, Copy)]
struct S;

impl A for S { fn f(self) {} }
impl B for S { fn f(self) {} }

fn main() {
    (&&&S).f(); // Here, we suggest `A::f((&&&S))` but we should suggest `A::f(***S)`
}

```

Doesn't need to be fixed in this PR just making note of it.